src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -90,33 +103,112 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
-  private List<Entity> handleCreated(String userEmail) {
+  private List<Entity> getHandleCreated(String userEmail) {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     List<Entity> results = new ArrayList<>();
 
     Query query =
         new Query("Event")
-            .setFilter(new Query.FilterPredicate("creator", Query.FilterOperator.EQUAL, userEmail))
-            .addSort("eventName", SortDirection.ASCENDING);
+            .setFilter(new Query.FilterPredicate("creator", Query.FilterOperator.EQUAL, userEmail));
+    // .addSort("eventName", SortDirection.ASCENDING);
     PreparedQuery queried = datastore.prepare(query);
     for (Entity e : queried.asIterable()) {
       results.add(e);
     }
-    // TODO: apply sort param choices
-
     return results;
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // adds or removes events from user's saved events list
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    UserService userService = UserServiceFactory.getUserService();
 
-    // TODO: add event to a list
+    if (userService.isUserLoggedIn()) {
+      String userEmail = userService.getCurrentUser().getEmail();
 
+      if (request.getParameter("event") != null) {
+        Long eventId = Long.parseLong(request.getParameter("event"));
+        Key userKey = KeyFactory.createKey("User", userEmail);
+
+        try {
+          Entity userEntity = datastore.get(userKey);
+          List<Long> saved = (ArrayList<Long>) userEntity.getProperty("saved");
+          if (saved == null) {
+            saved = new ArrayList<>();
+          }
+          switch (request.getParameter("action")) {
+            case "save":
+              postHandleSave(saved, eventId);
+              break;
+            case "unsave":
+              postHandleUnsave(saved, eventId);
+              break;
+            default:
+              throw new IOException("missing or invalid parameters");
+          }
+          userEntity.setProperty("saved", saved);
+          datastore.put(userEntity);
+        } catch (EntityNotFoundException exception) {
+          // datastore entry has not been created yet for this user, create it now
+          Entity entity = new Entity(userKey);
+          entity.setProperty("id", userEmail);
+          datastore.put(entity);
+        }
+
+      } else {
+        throw new IOException("no event key specified");
+      }
+    } else {
+      throw new IOException("must be logged in");
+    }
+  }
+
+  // adds event id to list if it is not already present
+  private void postHandleSave(List<Long> saved, long eventId) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Key eventKey = KeyFactory.createKey("Event", eventId);
+    try {
+      Entity eventEntity = datastore.get(eventKey);
+
+      for (int i = 0; i < saved.size(); i++) {
```

Comment:
Use `saved.contains(eventId)` instead of loop?

Reply:
Fixed!